File: server/src/domain/curved_grade_book.py
"""CurvedGradeBook — weighted grading with a configurable curve offset."""
from typing import List
from .grade_book import GradeBook, AssignmentScore, GradeResult


class CurvedGradeBook(GradeBook):
    def __init__(self, curve_offset: float = 5.0):
        self.curve_offset = curve_offset

    @property
    def scheme_name(self) -> str:
        return "curved"

    def calculate_grade(self, scores: List[AssignmentScore]) -> GradeResult:
        if not scores:
            return GradeResult(percentage=0.0, letter_grade="F", grade_points=0.0, display_label="F (Curved)")
        raw_pct = self._weighted_percentage(scores)
        curved_pct = min(100.0, raw_pct + self.curve_offset)
        letter = self._score_to_letter(curved_pct)
        return GradeResult(
            percentage=round(curved_pct * 10) / 10,
            letter_grade=letter,
            grade_points=self._letter_to_points(letter),
            display_label=f"{letter} (Curved +{int(self.curve_offset)}%)",
        )

File: artifacts/server/src/domain/CurvedGradeBook.ts
import { GradeBook, AssignmentScore, GradeResult } from "./GradeBook";

export class CurvedGradeBook extends GradeBook {
  constructor(private readonly curveOffset: number = 5) {
    super();
  }

  get schemeName(): string {
    return "curved";
  }

  calculateGrade(scores: AssignmentScore[]): GradeResult {
    if (scores.length === 0) {
      return { percentage: 0, letterGrade: "F", gradePoints: 0.0, displayLabel: "F (Curved)" };
    }
    const rawPct = this.weightedPercentage(scores);
    const curvedPct = Math.min(100, rawPct + this.curveOffset);
    const letter = this.scoreToLetter(curvedPct);
    return {
      percentage: Math.round(curvedPct * 10) / 10,
      letterGrade: letter,
      gradePoints: this.letterToPoints(letter),
      displayLabel: `${letter} (Curved +${this.curveOffset}%)`,
    };
  }
}
